### DISCUSSION TITLE
The ambiguity of `testClient()` documentation

### DISCUSSION THREAD

**Discussion by PlayerNguyen** [honojs/hono#2497] [General]
After searching a documentation for the endpoint testing, I found that the test helper `testClient` [documentation](https://hono.dev/helpers/testing) (which kinda depends on `hc<...>()`) is very ambigous. Since this discussion is created to 
- Improve the documentation by extending a test helper documentation
- Better TypeScript-ing support for `testClient<...>()` 

## Reason
After looking at the source code, I found that the function `testClient<>(app)` is returning a unknown type. Although I already tried to put the generic tags onto it.

## Examples

Here is my current project
```ts

// index.ts
const APIApp = new Hono()

// Main Application
APIApp.get("/test", (c) => c.text("success"));

// API Routes
const routes = APIApp.route("/a", ARouter)
  .route("/b", BRouter)
  

// export type APIAppRoutes = typeof routes;
// export default APIApp;

// endpoint/test.spec.ts
function getTestClientForApiApp() {  // this function is returning unknown type regardless trying to put any generic type into it
  return testClient<APIAppRoutes>(APIApp);
}

```

<img width="873" alt="image" src="https://github.com/honojs/hono/assets/10703461/f70b604c-7cdf-40e7-adee-56d136e5681c">

**Comment by PlayerNguyen** [comment]
Additional information
- Using Hono with Bun v1.0.30
- TypeScript (on package.json): 5.0.0
- VSCode, for sure.

**Comment by juniorgarcia** [comment]
Same was happening to me until I made some tests on how I create my Hono instance.

This works:
```.tsx
// app.spec.ts
import { testClient } from 'hono/testing'
import { Hono } from 'hono'

const app = new Hono().get('/', (c) =>
  c.json({
    message: 'Hello, Hono!',
  }),
)

describe('app', () => {
  it('works', async () => {
    const res = await testClient(app).index.$get()
    expect(await res.json()).toEqual({
      message: 'Hello, Hono!',
    })
  })
})
```

This does _not_:
```ts
// app.spec.ts
import { testClient } from 'hono/testing'
import { Hono } from 'hono'

const app = new Hono()
app.get('/', (c) =>
  c.json({
    message: 'Hello, Hono!',
  }),
)

describe('app', () => {
  it('works', async () => {
    const res = await testClient(app).index.$get() // Object is of type 'unknown'.ts(2571)
    expect(await res.json()).toEqual({
      message: 'Hello, Hono!',
    })
  })
})
```

So the problem is being caused by not chaining your routes directly after creating the `Hono` instance, making impossible for TypeScript to infer the type of the Hono client being created, which might not be the best approach since as far as I know (and I know very little about Hono) this might mess with middlewares since their order matters and some of them might not being applied for every situation.

**Reply by thenbe** [reply]
> So the problem is being caused by not chaining your routes directly after creating the Hono instance

I think you're correct. This section of the [docs](https://hono.dev/docs/guides/best-practices) mention the chaining part explicitly, and also offers an alternative workaround (`factory.createHandlers()`) that retains type safety.

I still haven't tried this myself yet though as I'm just researching the topic now.
